react-components/src/components/CadModelContainer/useApplyCadModelStyling.tsx

```diff
@@ -0,0 +1,125 @@
+            return;
+          }
+            
+          model.setDefaultNodeAppearance(DefaultNodeAppearance.Default);
+        };
+      }, [defaultStyle, model]);
+
+};
+
+async function applyStyling(
+    sdk: CogniteClient,
+    model: CogniteCadModel,
+    stylingGroups: Array<NodeStylingGroup | TreeIndexStylingGroup>,
+): Promise<void> {
+
+  const firstChangeIndex = getFirstChangeIndex();
+
+  for (let i = firstChangeIndex; i < model.styledNodeCollections.length; i++) {
+    const viewerStyledNodeCollection = model.styledNodeCollections[i];
+    model.unassignStyledNodeCollection(viewerStyledNodeCollection.nodeCollection);
+  }
+
+  for (let i = firstChangeIndex; i < stylingGroups.length; i++) {
+    const stylingGroup = stylingGroups[i];
+
+    if (stylingGroup.style === undefined)
+      continue;
+
+    if ('treeIndices' in stylingGroup) {
+      const nodes = new TreeIndexNodeCollection(stylingGroup.treeIndices); 
+      model.assignStyledNodeCollection(nodes, stylingGroup.style);
+    }
+    
+    if ('nodeIds' in stylingGroup) {
+      const nodes = new NodeIdNodeCollection(sdk, model);
+      await nodes.executeFilter(stylingGroup.nodeIds);
+      model.assignStyledNodeCollection(nodes, stylingGroup.style);
+    }
+  }
+
+  function getFirstChangeIndex(): number {
+    for (let i = 0; i < model.styledNodeCollections.length; i++) {
+      const stylingGroup = stylingGroups[i];
+      const viewerStyledNodeCollection = model.styledNodeCollections[i];
+
+      const areEqual = isEqualStylingGroupAndCollection(stylingGroup, viewerStyledNodeCollection);
+      
+      if (!areEqual) {
+        return i;
+      }
+    }
+
+    return model.styledNodeCollections.length;
+  };
+}
+
+function isEqualStylingGroupAndCollection(group: NodeStylingGroup | TreeIndexStylingGroup, collection: {
+  nodeCollection: NodeCollection;
+  appearance: NodeAppearance;
+}): boolean {
+  if (group?.style === undefined)
+    return false;
+
+  const isEqualGroupStyle = isEqualStyle(collection.appearance, group.style);
+
+  if (collection.nodeCollection instanceof TreeIndexNodeCollection && 'treeIndices' in group) {
+    const compareCollection = new TreeIndexNodeCollection(group.treeIndices)
+    const isEqualContent = isEqualTreeIndex(collection.nodeCollection, compareCollection);
+
+    return isEqualGroupStyle && isEqualContent;
+  }
+
+  if (collection.nodeCollection instanceof NodeIdNodeCollection && 'nodeIds' in group) {
+    const collectionNodeIds = collection.nodeCollection.serialize().state.nodeIds as Array<number>;
+    const isEqualContent = isEqual(collectionNodeIds, group.nodeIds);
+
+    return isEqualGroupStyle && isEqualContent;
+  }
+
+  return false;
+}
+
+function isEqualTreeIndex(collectionA: TreeIndexNodeCollection, collectionB: TreeIndexNodeCollection): boolean {
+  const isEqualContent = collectionA.getIndexSet().intersectWith(collectionB.getIndexSet()).count === collectionA.getIndexSet().count;
```

Comment:
Doesn't using the differenceWith make more sense here? I.e. `a.differenceWith(b).count === 0`? Or am I misunderstanding the semantics of `differenceWith`?

Also, the current implementation seems to be wrong for the case where B is a superset of A.
- A.intersect(B) = A
- A.count === A.count: true

But it should be false right?

Reply:
That's true, I forgot about this edge case because I thought that difference is done with more operations (union + intersection). Will use difference here.